staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/controller.go

```diff
@@ -0,0 +1,604 @@
+		}
+		if plName == fctypesv1a1.PriorityLevelConfigurationNameExempt && !haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll && !haveCatchAllPL {
+			klog.V(3).Infof("Retaining old priority level %q with Type=%v because of lack of replacement", plName, plState.config.Type)
+		} else {
+			if plState.queues == nil {
+				klog.V(3).Infof("Removing undesired priority level %q, Type=%v", plName, plState.config.Type)
+				continue
+			}
+			if plState.emptyHandler == nil {
+				klog.V(3).Infof("Priority level %q became undesired", plName)
+				plState.emptyHandler = &emptyRelay{cfgCtl: cfgCtl}
+				newlyQuiescent = append(newlyQuiescent, plState)
+			}
+		}
+		if plState.config.Limited != nil {
+			shareSum += float64(plState.config.Limited.AssuredConcurrencyShares)
+		}
+		haveExemptPL = haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameExempt
+		haveCatchAllPL = haveCatchAllPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll
+		newCfgState.priorityLevelStates[plName] = plState
+	}
+
+	// Supply missing mandatory objects
+	if !haveExemptPL {
+		newCfgState.imaginePL(fcboot.MandatoryPriorityLevelConfigurationExempt, cfgCtl.requestWaitLimit, &shareSum)
+	}
+	if !haveCatchAllPL {
+		newCfgState.imaginePL(fcboot.MandatoryPriorityLevelConfigurationCatchAll, cfgCtl.requestWaitLimit, &shareSum)
+	}
+
+	// For all the priority levels of the new config, divide up the
+	// server's total concurrency limit among them and create/update
+	// their QueueSets.
+	for plName, plState := range newCfgState.priorityLevelStates {
+		if plState.config.Limited == nil {
+			klog.V(5).Infof("Using exempt priority level %q: quiescent=%v", plName, plState.emptyHandler != nil)
+			continue
+		}
+
+		plState.qsConfig.ConcurrencyLimit = int(math.Ceil(float64(cfgCtl.serverConcurrencyLimit) * float64(plState.config.Limited.AssuredConcurrencyShares) / shareSum))
+		metrics.UpdateSharedConcurrencyLimit(plName, plState.qsConfig.ConcurrencyLimit)
+
+		if plState.queues == nil {
+			klog.V(5).Infof("Introducing queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescent=%v (shares=%v, shareSum=%v)", plName, plState.config, plState.qsConfig.ConcurrencyLimit, plState.emptyHandler != nil, plState.config.Limited.AssuredConcurrencyShares, shareSum)
+			plState.queues = cfgCtl.queueSetFactory.NewQueueSet(*plState.qsConfig)
+		} else {
+			klog.V(5).Infof("Retaining queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescent=%v (shares=%v, shareSum=%v)", plName, plState.config, plState.qsConfig.ConcurrencyLimit, plState.emptyHandler != nil, plState.config.Limited.AssuredConcurrencyShares, shareSum)
+			plState.queues.SetConfiguration(*plState.qsConfig)
+		}
+	}
+
+	// The new config has been constructed, pass to filter for use.
+	cfgCtl.curState.Store(newCfgState)
+	klog.V(5).Infof("Switched to new RequestManagementState")
+
+	// We delay the calls to QueueSet::Quiesce so that the following
```

Comment:
Let me try to restate what @jktomer said.

The problem with this isn't that one can't generate the proof--obviously you did. The problem is that anyone making changes in the file, or reviewing the file, needs to be able to mentally generate a similar proof for every line they change or review.

This is not a simple proof. Yes, it is short, but that doesn't make it simple. "Simple", in my vocabulary, means that it is easy to *find*, not that it is easy to *state*. The difficulty of finding a particular proof goes with the space one has to search to find it. So, proofs that imply a large search space are complex in my vocabulary. Search space increases with distance in the codebase, and combinatorially with the number of terms in the proof.